# **Analytical Comparison Between CEO Concerns**


> ## _Author: Ronald Washington III_ 

> > Our society is currently run by some of the worlds largest and smartest technology
companies. Where each of these companies have a vision and impact on the world due to the
views of their CEO. But how do the views between these CEOs differ? What are the topics that
these CEOs are constantly concerned with and are expressing their opinions? Within this study
we shall collect tweets from selected CEOs, determine if we are able to identify the tweet’s
originator correctly, and extract the topic of each tweet to have a collection of concerns that each
CEO has. From there the analysis can be furthered by calculating the distance between the issues
that the CEOs talk about and have an understanding of how closely related their concerns are.

> > Conducting our analysis through the assistance of an classification algorithm will aid in
our research concerning the topics that CEOs primarily tweet about. The specific algorithm that
will be performed in this research will be Support Vector Machines, which is known to be useful
in the classification of images, bioinformatics, and text categorization. Support Vector Machines
(SVMs) are going to be useful determining which CEOs a tweet belong to, and we will be
utilizing Latent Dirichlet Allocation to extract prominent topics among the classified tweets.
Through the usage of both these techniques, we strive to accomplish our goal of classifying
tweets and finding the common topics discussed among CEOs. By the end of this research we
shall attempt to distinguish the difference between topics of CEOs, the similarities/differences,
and the number of times a specific topic is tweeted.


### Imported Libraries 

In [1]:
import keys as k
import pandas as pd
import webbrowser
import requests
import bs4
import importlib
import tweepy
import json 
from requests_oauthlib import OAuth1Session
importlib.reload(k)

<module 'keys' from 'C:\\Users\\Bloody Dachi\\Documents\\CS_401\\Final_Project\\CEO-Topic-Classifier-Extractor\\keys.py'>

### Preparing/Gathering Authorization for Twitter Access

In [2]:
twitter = k.twitter
client_key = twitter["client_key"]
client_secret = twitter["client_secret"]
resource_owner_key = twitter['resource_owner_key']
resource_owner_secret = twitter['resource_owner_secret']
protected_url = 'https://api.twitter.com/1.1/statuses/user_timeline.json?screen_name=jlist'

session = OAuth1Session(client_key,
                        client_secret=client_secret,
                        resource_owner_key=resource_owner_key,
                        resource_owner_secret= resource_owner_secret)

consumer_key = twitter["client_key"]
consumer_secret = twitter["client_secret"]
access_token = twitter['resource_owner_key']
access_token_secret =  twitter['resource_owner_secret']
auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_token_secret)
api = tweepy.API(auth,wait_on_rate_limit=True)

# Developing a list of CEOs

> Within this study we will be utlizing a list of CEOs provided by Valeria Maltoni (@ConversationAge), who created a member group/list of CEOs who use Twitter. 
- https://twitter.com/ConversationAge/lists/ceos/members?lang=en

In [3]:

group_name = 'ceos'
owner = 'ConversationAge'


def get_list_members(api, owner, group_name):
    members = []
    # without this you only get the first 20 list members
    for person in tweepy.Cursor(api.list_members, owner, group_name).items():
        number_tweets = person.statuses_count
        if number_tweets>=1000:
            members.append(person)
    # create a list containing all usernames
    return [ m.screen_name for m in members ]

In [4]:
len(get_list_members(api,owner,group_name))

655

In [5]:
CEO_list = pd.DataFrame()
CEO_list['CEO_Twitter_Accounts'] = get_list_members(api,owner,group_name)
CEO_list.head(5)

CEO_Twitter_Accounts
0         DrJudyMonroe
1       JillStorey2020
2           co_opcloud
3           PaulPolman
4            KenLingad

In [6]:
len(CEO_list)

655

In [7]:
PrivateList = []
TweetCountList = []
import time
t0 = time.time()

for i in range(len(CEO_list)):
    number_tweets = api.get_user(CEO_list['CEO_Twitter_Accounts'][i]).statuses_count
    if number_tweets>=1000:
        x = api.get_user(CEO_list['CEO_Twitter_Accounts'][i])
        TweetCountList.append(number_tweets)
        PrivateList.append(x.protected)
    #print(x.screen_name,x.protected)
t1 = time.time()
total = ((t1-t0)/60)
print("Time to Run: ", total)
import collections
counter=collections.Counter(PrivateList)
print("Count of Non-Private Versus Private Twitter Accounts: ", counter)

Time to Run:  16.520678122838337
Count of Non-Private Versus Private Twitter Accounts:  Counter({False: 651, True: 4})


In [8]:
se = pd.Series(PrivateList)
te = pd.Series(TweetCountList)
CEO_list['Private_Indicator'] = se.values
CEO_list['Tweet_Count'] = te.values

In [9]:
CEO_list[25:30]

CEO_Twitter_Accounts  Private_Indicator  Tweet_Count
25            techUKCEO              False         3274
26         morganberman              False         1553
27           jane_knows              False         1373
28             limouris              False         2192
29        Monique_Villa              False         1532

In [10]:
CEO_list_cleaned = pd.DataFrame()
CEO_list_cleaned = CEO_list
CEO_list_cleaned = CEO_list_cleaned[CEO_list['Private_Indicator']!=True]

In [11]:
len(CEO_list_cleaned)

651

In [12]:
CEO_list_cleaned[24:222]

CEO_Twitter_Accounts  Private_Indicator  Tweet_Count
24            HaniSFarsi              False         1109
25             techUKCEO              False         3274
26          morganberman              False         1553
27            jane_knows              False         1373
28              limouris              False         2192
29         Monique_Villa              False         1532
30        DelMonteLouis1              False         3639
31           AbdelQaader              False        29153
32         CarolynAHardy              False         9181
33            heylizelle              False        19947
34        AnthonyFerrier              False         1815
35            IDollyLenz              False        39501
36            EllenAgler              False         4071
37         janemariechen              False         1037
38          CarrieMantha              False         7208
39            TerriSeese              False         1219
40          scottpetinga              False         2163
41          DebraMJacobs              False        24349
42             ZenYinger              False        41828
43            BrookBello              False         7017
44          nickstanhope              False         2208
45             Bill_Nuti              False         1381
46            katrynadow              False        13006
47           YaelLehmann              False         2967
48            zackkanter              False         7808
49          RachelGerrol              False         3050
50       HowHackersThink              False         6558
51             gbaileyco              False        11661
52         CherryStreet7              False         6988
53            BobRuffolo              False        19813
..                   ...                ...          ...
196             dancathy              False         3253
197          kristinhull              False         4940
198        craigwortmann              False         4839
199          M_CarolineH              False         2579
200       WalterIsaacson              False         2608
201      joshmccall_jack              False         1239
202           milesnadal              False        33003
203          HauteLiving              False         4882
204      ManpowerGroupJJ              False         2718
205          FranckOhrel              False        14983
206         ericschiffer              False         2041
207         Tommy_Cannon              False         9435
208          FredCuellar              False        57745
209          WDavidPower              False         1724
210          AnitaGrover              False         1061
211              sklasko              False         2801
212         robertjmoore              False         3143
213      simonmainwaring              False        38902
214             JoshuaRM              False         2194
215           TerenaBell              False        19887
216            mddelphis              False        20058
217        emilymiethner              False        15967
218          Bill_Simmel              False        41070
219        aaron_epstein              False         4388
220       michaelchatman              False        66853
221           justinwisz              False         2113
222       jposhaughnessy              False        24251
223           lucymarcus              False        45354
224        robertucraven              False         5047
225         jacobcharold              False         4685

[198 rows x 3 columns]

# Gathering Tweets

In [19]:
import tweepy
import json
import time
def tweetCollector(dataframe):
    """
    Creates a CSV for every CEO's tweets 
    """
    function_start = time.time()
    list_tweets = []
    name = []
    user_name = []
    count = 0
    for i in dataframe:
        
        for page in tweepy.Cursor(api.user_timeline, screen_name='@'+i,include_rts = False, tweet_mode = 'extended').pages(100):
            for status in page:

                list_tweets.append(status._json['full_text'])
                name.append(status._json['user']['name'])
                user_name.append(status._json['user']['screen_name'])
                DataSet = pd.DataFrame()

                DataSet['Tweet'] = [tweet for tweet in list_tweets]
                DataSet['CEO_Full_Name'] = [tweet for tweet in name]
                DataSet['CEO_User_Name'] = [tweet for tweet in user_name]

    #either one CSV for all the CEO
        count +=1
        print("CEO #: ", count, "CEO Name: ",i)
    DataSet.to_csv("Collection_CEO_TWEETS.csv")
    
    function_end = time.time()
    total_function_time = ((function_end-function_start)/60)
    print("Time to Complete: ", total_function_time)
    
    return DataSet
#tweetCollector(CEO_list_cleaned['CEO_Twitter_Accounts'][:50]) 1
#tweetCollector(CEO_list_cleaned['CEO_Twitter_Accounts'][50:99])2
#tweetCollector(CEO_list_cleaned['CEO_Twitter_Accounts'][99:149])3
#tweetCollector(CEO_list_cleaned['CEO_Twitter_Accounts'][149:199])4
#tweetCollector(CEO_list_cleaned['CEO_Twitter_Accounts'][199:249])5 
#tweetCollector(CEO_list_cleaned['CEO_Twitter_Accounts'][249:299])6
#tweetCollector(CEO_list_cleaned['CEO_Twitter_Accounts'][299:349])7
#tweetCollector(CEO_list_cleaned['CEO_Twitter_Accounts'][349:399])8
#tweetCollector(CEO_list_cleaned['CEO_Twitter_Accounts'][399:449])9
#tweetCollector(CEO_list_cleaned['CEO_Twitter_Accounts'][449:499])10
#tweetCollector(CEO_list_cleaned['CEO_Twitter_Accounts'][499:549])11
#tweetCollector(CEO_list_cleaned['CEO_Twitter_Accounts'][549:599])12
#tweetCollector(CEO_list_cleaned['CEO_Twitter_Accounts'][599:])13

CEO #:  1 CEO Name:  ScottMonty
CEO #:  2 CEO Name:  levie
CEO #:  3 CEO Name:  mukund
CEO #:  4 CEO Name:  peretti
CEO #:  5 CEO Name:  thornley
CEO #:  6 CEO Name:  jeckman
CEO #:  7 CEO Name:  kduggan
CEO #:  8 CEO Name:  thinkfuture
CEO #:  9 CEO Name:  paulmcenany
CEO #:  10 CEO Name:  FrankGruber
CEO #:  11 CEO Name:  stuartma
CEO #:  12 CEO Name:  marshallk
CEO #:  13 CEO Name:  davidcohen
CEO #:  14 CEO Name:  Huperniketes
CEO #:  15 CEO Name:  jeremywright
CEO #:  16 CEO Name:  gianandrea
CEO #:  17 CEO Name:  BizGirl
CEO #:  18 CEO Name:  aripap
CEO #:  19 CEO Name:  webwright
CEO #:  20 CEO Name:  zsims
CEO #:  21 CEO Name:  ceonyc
CEO #:  22 CEO Name:  johnolilly
CEO #:  23 CEO Name:  carolyn
CEO #:  24 CEO Name:  loic
CEO #:  25 CEO Name:  gc
CEO #:  26 CEO Name:  TDefren
CEO #:  27 CEO Name:  kevinrose
CEO #:  28 CEO Name:  chaddickerson
CEO #:  29 CEO Name:  andybudd
CEO #:  30 CEO Name:  anildash
CEO #:  31 CEO Name:  ryancarson
CEO #:  32 CEO Name:  photomatt
CEO #:  3

Tweet CEO_Full_Name  \
0      @MarkWarner @KrisColvin Less tweeting. More le...   Scott Monty   
1                        @cc_chapman @evoterra Flipboard   Scott Monty   
2      Grateful to @Anchor for making it possible for...   Scott Monty   
3      @MattNavarra Sheesh. Can't get it on Tumblr, c...   Scott Monty   
4              That link, tho... https://t.co/cTso8z3l2t   Scott Monty   
5      @renato_mariotti @justinhendrix There's a good...   Scott Monty   
6      @KyleRohde Well, since he has me blocked, this...   Scott Monty   
7                    @nathanbarry @RealAndyJackson Whew.   Scott Monty   
8      New from Pew Research: Americans Still Prefer ...   Scott Monty   
9      @nathanbarry @RealAndyJackson Hmm, that’s weir...   Scott Monty   
10     @RealAndyJackson @nathanbarry Thanks Andy!\n\n...   Scott Monty   
11     @jayacunzo How about Bad Science, where a come...   Scott Monty   
12     Four options to transcend the social media clu...   Scott Monty   
13     This week's #FullMonty includes a discussion o...   Scott Monty   
14     @bhrome @jayacunzo I give people plenty to fin...   Scott Monty   
15                              @mkonnikova That's nuts.   Scott Monty   
16     Some Timeless Wisdom from Seneca on this #Moti...   Scott Monty   
17     @Dvrbkwrm Thanks so much, Carolyn. I’m sure Ci...   Scott Monty   
18                       @MikeVJennings Thank you, Mike!   Scott Monty   
19     The Human Fund — December 3, 2018 https://t.co...   Scott Monty   
20     New from Pew Research: Americans unhappy with ...   Scott Monty   
21     From farnamstreet: Love, Happiness, and Time h...   Scott Monty   
22     @douglaskarr That's a heck of a stool sample y...   Scott Monty   
23     @ChuckWendig @KyleRohde We've got Echo Dot on ...   Scott Monty   
24     @jayacunzo As referenced in:\nhttps://t.co/l4K...   Scott Monty   
25     @jayacunzo "There is no more stupefying thing ...   Scott Monty   
26       He should have sent it. https://t.co/vHqCP45PbM   Scott Monty   
27     @sarahfrier @profgalloway Check out Celebritie...   Scott Monty   
28     @KLemkau Remember when we used to vlog from ho...   Scott Monty   
29     @anchor Sent an inquiry to you guys about dist...   Scott Monty   
...                                                  ...           ...   
78589  Starting now #BlackPantherLive https://t.co/Vu...          jack   
78590                  @roach Feels like a bug @kcoleman          jack   
78591         #BlackPantherLive  https://t.co/3JLC5V2HmK          jack   
78592                  Love this https://t.co/2JngyX3zih          jack   
78593  Tomorrow: #BlackPantherLive https://t.co/Vu58c...          jack   
78594                                         @dplenge K          jack   
78595                                     @MikeIsaac Yes          jack   
78596                               @Jury Sounds amazing          jack   
78597                            https://t.co/HtHJxYjNdK          jack   
78598                  @thecleanmachine Literally my fav          jack   
78599                   @thecleanmachine Best place ever          jack   
78600                                       @Square Yes.          jack   
78601  9 years of being perfectly Square! https://t.c...          jack   
78602  Can’t get over how brilliant this is @kendrick...          jack   
78603  All The Stars @kendricklamar @sza https://t.co...          jack   
78604  @leslieberland @heySierra @omidkordestani @bru...          jack   
78605  I’m so proud of the @Twitter team. We did what...          jack   
78606                            https://t.co/5dis2Q96yL          jack   
78607                                 @AnaBotin Welcome!          jack   
78608                            https://t.co/IiygHKYv7U          jack   
78609      Access all the tweets https://t.co/3ioIJuwHyM          jack   
78610  Rotate your phone around when you read the boo...          jack   
78611  Instant buying (and selling, if you don’t want..

- Gathered threes from 651 CEOs
- Put all tweets together into a large corpus (Each CEO gets <1000 tweets...combined 907k tweets roughly)
- Perform LDA to do topic extraction from a selected CEOs tweet corpus, then do the same but for all CEOs
- Attempt performing Support Vector Machine Algorithm to classify which tweet belongs to which CEO